server/backend/cron/news.go

```diff
@@ -131,53 +135,90 @@ func (c *CronService) parseNewsFeed(source model.NewsSource) error {
 	return nil
 }
 
-// downloadAndSaveImage
-// Downloads an image from the url, converts it to 720p width to and saves it to the database.
-// Returns id of file in db, null int on error.
-func (c *CronService) downloadAndSaveImage(url string) null.Int {
+// getDatabaseIdForImageAndDownload
+// Returns a file id that was saved to the database if it already exists. Otherwise creates it and triggers download of the image
+func (c *CronService) getDatabaseIdForImageAndDownload(url string) (null.Int, error) {
+	targetFileName := fmt.Sprintf("%x.jpg", md5.Sum([]byte(url)))
+	var fileId null.Int
+	if err := c.db.Model(model.Files{}).Where("name = ?", targetFileName).Select("file").Scan(&fileId).Error; err != nil && err != gorm.ErrRecordNotFound {
+		log.Printf("Couldn't query database for file: %v", err)
+		return null.Int{}, err
+	}
+	if fileId.Valid { // file already in database -> return for current news.
+		return fileId, nil
+	}
+
+	// otherwise store in database:
+	file := model.Files{Name: targetFileName, Path: STORAGE_DIR + ImageDirectory}
+	if err := c.db.Transaction(func(tx *gorm.DB) error {
+		if err := tx.Create(&file).Error; err != nil {
+			log.Printf("Could not store new file to database: %v", err)
+			return err
+		}
+		return nil
+	}); err != nil {
+		return null.Int{}, err
+	}
+
+	go c.downloadFile(url, targetFileName, MAX_IMAGE_RETRYS)
+	return null.Int{NullInt64: sql.NullInt64{Valid: true, Int64: int64(file.File)}}, nil
+}
+
+// downloadFile tries downloading a file errorCounter times. After the errorCounters failure the corresponding entry is deleted from the database.
+//
+// url: download url of the file
+// name: target name of the file
+// errorCounter: recursively decremented counter for errors.
+func (c *CronService) downloadFile(url string, name string, errorCounter int) {
+	log.Printf("downloading file %s", url)
+	if errorCounter == 0 {
+		// unable to download image. Delete from database.
+		var newsWithBadFile []model.News
+		if err := c.db.Model(&model.News{}).Joins("JOIN files on files.file = news.file WHERE files.name = ?", name).Scan(&newsWithBadFile).Error; err != nil && err != gorm.ErrRecordNotFound {
+			log.Println("Could not get news with bad files: %v", err)
+			sentry.CaptureException(err)
+			return
+		}
+		for i := range newsWithBadFile {
+			newsWithBadFile[i].File = null.Int{NullInt64: sql.NullInt64{Valid: false}}
+			if err := c.db.Save(&newsWithBadFile).Error; err != nil {
```

Comment:
I'd probably just would delete the news entry as well and let it be re-added with the next rss parsing. Unless our RSS parsing logic notices missing files, we'd never try the same download again right?